lib/web_ui/dev/webdriver_browser.dart

```diff
@@ -0,0 +1,96 @@
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+import 'dart:math';
+
+import 'package:image/image.dart';
+import 'package:webdriver/io.dart' show createDriver, WebDriver;
+
+import 'browser.dart';
+
+abstract class WebDriverBrowserEnvironment extends BrowserEnvironment {
+  late final Process _driverProcess;
+
+  Future<Process> spawnDriverProcess();
+  Uri get driverUri;
+
+  @override
+  Future<void> prepare() async {
+    _driverProcess = await spawnDriverProcess();
+
+    _driverProcess.stderr
+        .transform(utf8.decoder)
+        .transform(const LineSplitter())
+        .listen((String error) {
+      print('[Webdriver][Error] $error');
+    });
+
+    _driverProcess.stdout
+        .transform(utf8.decoder)
+        .transform(const LineSplitter())
+        .listen((String log) {
+      print('[Webdriver] $log');
+    });
+  }
+
+  @override
+  Future<void> cleanup() async {
+    _driverProcess.kill();
+  }
+
+  @override
+  Future<Browser> launchBrowserInstance(Uri url, {bool debug = false}) async {
+    for (;;) {
+      try {
+        final WebDriver driver = await createDriver(
+          uri: driverUri, desired: <String, dynamic>{'browserName': packageTestRuntime.identifier});
+        return WebDriverBrowser(driver, url);
+      } on SocketException {
```

Comment:
I wonder if it would be more appropriate to smoke test (and retry if necessary) the driver process in the `prepare` method. This way other methods that need the driver to be ready don't need to mistrust the initialization process of the driver.

Reply:
I tried coming up with a good solution here based on your suggestion, but I'm not really sure I really like anything I could come up with. I understand where you're coming from in that it seems like the responsibility of "prepare" to ensure that the web driver process is actually ready to go, but I think actually trying to enforce this only ends up with negatives practically speaking. For one thing, doing the smoke test in `prepare` results in opening an extra socket just to check if the driver process is ready, which is a non-trivial operation from a perf perspective. In practice, this is being validated the first time `launchBrowserInstance` is invoked, and after that, other browser launches should end up in the happy path. The only real benefit I could see to doing the validation in prepare is that if there is some situation where the web driver launches and the smoke test passes, but something bad happens in the web driver process before a subsequent invocation of `launchBrowserInstance` then we would fail fast rather than have to wait for the test timeout. But this seems like an edge case to optimize for that (hopefully) wouldn't happen too often in practice anyway.